File: hub/store/address.ts
import { HubError } from "./errors.ts";

const RESERVED_AGENT_SLUGS = new Set(["default", "all"]);
const AGENT_SLUG_RE = /^[a-z0-9-]{1,32}$/;

/** How a display address should be resolved. */
export type AddressKind = "user" | "org_broadcast" | "self_agent" | "my_agents";

export interface ParsedDisplayAddress {
  kind: AddressKind;
  local: string;
  orgSlug: string;
  agentSlug?: string;
}

/** Org-wide broadcast: `@all@org`. */
export function isBroadcastHandle(handle: string): boolean {
  return handle.startsWith("@all@");
}

/** Self fan-out to all of the current user's agents: bare `@all`. */
export function isMyAgentsHandle(handle: string): boolean {
  return handle.trim() === "@all";
}

export function broadcastHandle(orgSlug: string): string {
  return `@all@${orgSlug}`;
}

export function myAgentsHandle(): string {
  return "@all";
}

export function assertValidAgentSlug(slug: string): void {
  if (!AGENT_SLUG_RE.test(slug)) {
    throw new HubError(
      "Agent slug must be 1–32 lowercase letters, digits, or hyphens",
      "invalid_agent_slug",
    );
  }
  if (RESERVED_AGENT_SLUGS.has(slug)) {
    throw new HubError(`Agent slug '${slug}' is reserved`, "invalid_agent_slug");
  }
}

/**
 * Parse display addresses:
 * - `alice@acme`, `alice@acme/claude`
 * - `@all@acme` (org broadcast)
 * - `@all` (all of my agents)
 * - `@claude` (shorthand for current user's agent slug)
 */
export function parseDisplayAddress(input: string): ParsedDisplayAddress {
  const trimmed = input.trim();

  if (isMyAgentsHandle(trimmed)) {
    return { kind: "my_agents", local: "@all", orgSlug: "", agentSlug: undefined };
  }

  if (isBroadcastHandle(trimmed)) {
    const orgSlug = trimmed.slice("@all@".length);
    if (!orgSlug) {
      throw new HubError("Invalid broadcast handle", "invalid_handle");
    }
    return { kind: "org_broadcast", local: "@all", orgSlug, agentSlug: undefined };
  }

  // Self-agent shorthand: `@claude` (single leading @, no other @ or /).
  if (
    trimmed.startsWith("@") &&
    !trimmed.includes("/", 1) &&
    trimmed.indexOf("@", 1) === -1
  ) {
    const slug = trimmed.slice(1);
    if (!slug) throw new HubError("Invalid handle format", "invalid_handle");
    assertValidAgentSlug(slug);
    return { kind: "self_agent", local: "", orgSlug: "", agentSlug: slug };
  }

  const slash = trimmed.indexOf("/");
  const base = slash >= 0 ? trimmed.slice(0, slash) : trimmed;
  const agentSlug = slash >= 0 ? trimmed.slice(slash + 1) : undefined;
  if (agentSlug !== undefined) {
    if (!agentSlug) throw new HubError("Missing agent slug after /", "invalid_handle");
    assertValidAgentSlug(agentSlug);
  }

  const at = base.lastIndexOf("@");
  if (at <= 0 || at === base.length - 1) {
    throw new HubError("Invalid handle format", "invalid_handle");
  }
  const local = base.slice(0, at);
  const orgSlug = base.slice(at + 1);
  if (local.toLowerCase() === "@all" || local.toLowerCase().startsWith("@all")) {
    throw new HubError("Handle cannot use @all broadcast prefix", "invalid_handle");
  }
  return { kind: "user", local, orgSlug, agentSlug };
}

export function bareHandle(local: string, orgSlug: string): string {
  return `${local}@${orgSlug}`;
}

export function formatDisplayAddress(
  local: string,
  orgSlug: string,
  agentSlug?: string,
): string {
  const base = bareHandle(local, orgSlug);
  return agentSlug ? `${base}/${agentSlug}` : base;
}

export function stripAgentSuffix(display: string): string {
  const slash = display.indexOf("/");
  return slash >= 0 ? display.slice(0, slash) : display;
}

export function agentSuffix(display: string): string | undefined {
  const slash = display.indexOf("/");
  return slash >= 0 ? display.slice(slash + 1) : undefined;
}

/** Wire path: `acme/alice/claude` (org/user/agent). */
export function formatWirePath(orgSlug: string, local: string, agentSlug: string): string {
  return `${orgSlug}/${local}/${agentSlug}`;
}

export function parseWirePath(path: string): { orgSlug: string; local: string; agentSlug: string } {
  const parts = path.split("/");
  if (parts.length !== 3 || !parts[0] || !parts[1] || !parts[2]) {
    throw new HubError("Invalid wire path (expected org/user/agent)", "invalid_handle");
  }
  assertValidAgentSlug(parts[2]);
  return { orgSlug: parts[0], local: parts[1], agentSlug: parts[2] };
}

export function assertHandleLocal(local: string): void {
  if (local.toLowerCase() === "@all" || local.toLowerCase().startsWith("@all")) {
    throw new HubError("Handle cannot use @all broadcast prefix", "invalid_handle");
  }
}

/** Legacy user handle parse (no agent suffix). Rejects shorthand / my-agents forms. */
export function parseUserHandle(handle: string): { local: string; orgSlug: string } {
  const parsed = parseDisplayAddress(handle);
  if (parsed.kind !== "user" && parsed.kind !== "org_broadcast") {
    throw new HubError("Invalid handle format", "invalid_handle");
  }
  return { local: parsed.local, orgSlug: parsed.orgSlug };
}

File: core/src/address.rs
//! Display ↔ wire address transforms for agent-scoped handles.

use std::collections::HashSet;
use std::sync::LazyLock;

use anyhow::{Context, Result, bail};

static RESERVED: LazyLock<HashSet<&'static str>> =
    LazyLock::new(|| HashSet::from(["default", "all"]));

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum AddressKind {
    User,
    OrgBroadcast,
    SelfAgent,
    MyAgents,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct ParsedDisplayAddress {
    pub kind: AddressKind,
    pub local: String,
    pub org_slug: String,
    pub agent_slug: Option<String>,
}

/// Org-wide broadcast: `@all@org`.
pub fn is_broadcast_handle(handle: &str) -> bool {
    handle.starts_with("@all@")
}

/// Shared my-agents group: bare `@all`.
pub fn is_my_agents_handle(handle: &str) -> bool {
    handle.trim() == "@all"
}

pub fn broadcast_handle(org_slug: &str) -> String {
    format!("@all@{org_slug}")
}

pub fn my_agents_handle() -> &'static str {
    "@all"
}

fn is_valid_agent_slug(slug: &str) -> bool {
    if slug.is_empty() || slug.len() > 32 {
        return false;
    }
    slug.bytes()
        .all(|b| b.is_ascii_lowercase() || b.is_ascii_digit() || b == b'-')
}

pub fn assert_valid_agent_slug(slug: &str) -> Result<()> {
    if !is_valid_agent_slug(slug) {
        bail!("invalid agent slug: must be 1–32 lowercase letters, digits, or hyphens");
    }
    if RESERVED.contains(slug) {
        bail!("agent slug '{slug}' is reserved");
    }
    Ok(())
}

/// Parse `alice@acme`, `alice@acme/claude`, `@all@acme`, bare `@all`, or `@claude`.
pub fn parse_display_address(input: &str) -> Result<ParsedDisplayAddress> {
    let trimmed = input.trim();

    if is_my_agents_handle(trimmed) {
        return Ok(ParsedDisplayAddress {
            kind: AddressKind::MyAgents,
            local: "@all".into(),
            org_slug: String::new(),
            agent_slug: None,
        });
    }

    if is_broadcast_handle(trimmed) {
        let org_slug = trimmed.strip_prefix("@all@").unwrap_or("");
        if org_slug.is_empty() {
            bail!("invalid broadcast handle");
        }
        return Ok(ParsedDisplayAddress {
            kind: AddressKind::OrgBroadcast,
            local: "@all".into(),
            org_slug: org_slug.into(),
            agent_slug: None,
        });
    }

    // Self-agent shorthand: `@claude` (single leading @, no other @ or /).
    if trimmed.starts_with('@')
        && !trimmed[1..].contains('/')
        && !trimmed[1..].contains('@')
    {
        let slug = &trimmed[1..];
        if slug.is_empty() {
            bail!("invalid handle format");
        }
        assert_valid_agent_slug(slug)?;
        return Ok(ParsedDisplayAddress {
            kind: AddressKind::SelfAgent,
            local: String::new(),
            org_slug: String::new(),
            agent_slug: Some(slug.to_string()),
        });
    }

    let (base, agent_slug) = match trimmed.split_once('/') {
        Some((b, slug)) => {
            if slug.is_empty() {
                bail!("missing agent slug after /");
            }
            assert_valid_agent_slug(slug)?;
            (b, Some(slug.to_string()))
        }
        None => (trimmed, None),
    };

    let at = base.rfind('@').context("invalid handle format")?;
    if at == 0 || at == base.len() - 1 {
        bail!("invalid handle format");
    }
    let local = base[..at].to_string();
    let org_slug = base[at + 1..].to_string();
    if local.eq_ignore_ascii_case("@all") || local.to_ascii_lowercase().starts_with("@all") {
        bail!("handle cannot use @all broadcast prefix");
    }

    Ok(ParsedDisplayAddress {
        kind: AddressKind::User,
        local,
        org_slug,
        agent_slug,
    })
}

pub fn bare_handle(local: &str, org_slug: &str) -> String {
    format!("{local}@{org_slug}")
}

pub fn format_display_address(local: &str, org_slug: &str, agent_slug: Option<&str>) -> String {
    match agent_slug {
        Some(slug) => format!("{local}@{org_slug}/{slug}"),
        None => bare_handle(local, org_slug),
    }
}

pub fn strip_agent_suffix(display: &str) -> &str {
    display.split('/').next().unwrap_or(display)
}

pub fn agent_suffix(display: &str) -> Option<&str> {
    display.split_once('/').map(|(_, slug)| slug)
}

pub fn format_wire_path(org_slug: &str, local: &str, agent_slug: &str) -> String {
    format!("{org_slug}/{local}/{agent_slug}")
}

pub fn parse_wire_path(path: &str) -> Result<(String, String, String)> {
    let parts: Vec<&str> = path.split('/').collect();
    if parts.len() != 3 || parts.iter().any(|p| p.is_empty()) {
        bail!("invalid wire path (expected org/user/agent)");
    }
    assert_valid_agent_slug(parts[2])?;
    Ok((parts[0].into(), parts[1].into(), parts[2].into()))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parse_and_format_roundtrip() {
        let p = parse_display_address("alice@acme/claude").unwrap();
        assert_eq!(p.kind, AddressKind::User);
        assert_eq!(p.local, "alice");
        assert_eq!(p.org_slug, "acme");
        assert_eq!(p.agent_slug.as_deref(), Some("claude"));
        assert_eq!(
            format_display_address(&p.local, &p.org_slug, p.agent_slug.as_deref()),
            "alice@acme/claude"
        );
        assert_eq!(format_wire_path("acme", "alice", "claude"), "acme/alice/claude");
    }

    #[test]
    fn parse_self_agent_shorthand() {
        let p = parse_display_address("@claude").unwrap();
        assert_eq!(p.kind, AddressKind::SelfAgent);
        assert_eq!(p.agent_slug.as_deref(), Some("claude"));
    }

    #[test]
    fn parse_my_agents_vs_org_broadcast() {
        let mine = parse_display_address("@all").unwrap();
        assert_eq!(mine.kind, AddressKind::MyAgents);
        let org = parse_display_address("@all@acme").unwrap();
        assert_eq!(org.kind, AddressKind::OrgBroadcast);
        assert_eq!(org.org_slug, "acme");
        assert!(is_my_agents_handle("@all"));
        assert!(!is_my_agents_handle("@all@acme"));
        assert!(is_broadcast_handle("@all@acme"));
        assert!(!is_broadcast_handle("@all"));
    }

    #[test]
    fn strip_suffix() {
        assert_eq!(strip_agent_suffix("bob@acme/research"), "bob@acme");
        assert_eq!(agent_suffix("bob@acme/research"), Some("research"));
    }

    #[test]
    fn reserved_slug_rejected() {
        assert!(parse_display_address("x@y/default").is_err());
        assert!(parse_display_address("@default").is_err());
        assert!(parse_display_address("@all").is_ok()); // my_agents, not slug
    }
}
